### PR TITLE
Add tail-free and top-a sampling

### PR DESCRIPTION
**Pull request by toast22a**
This PR adds a provision for adding custom samplers when using the `transformers` library for generation. We include the tail-free and top-a samplers, but the file `sampler_hijack.py` can be extended later to add more samplers.

To do this, we monkey-patch:
* `transformers.GenerationMixin` to add the new samplers into `_get_logits_warper()` after the existing samplers, but before logit normalization
* `transformers.GenerationConfig` to add the new samplers' values as generation arguments

This is intended to be a temporary solution to #2198 until HF adds these samplers to the `transformers` library. When that happens, it is trivial to just remove the custom samplers from `sampler_hijack.py`.

### CODE REVIEW

**Comment by BadisG** [comment]
I have the feeling that if you change the value of one of those parameters, it doesn't change anything unless you're doing a second generation.

**Comment by oobabooga** [comment]
I appreciate the effort to provide new samplers, but I don't think it's worth it to add a new monkey patch just for this. A similar parameter already exists in the ui (eta_sampling). I have recently added a proper preset for using it: https://github.com/oobabooga/text-generation-webui/commit/6de727c5242fc9933260bfd45f35d8310eba4a57

**Comment by oobabooga** [comment]
Well, ffs, so many people are asking for this that I'll just merge it before I get lynched
